### ISSUE TITLE
[performance] Remove usage of get-in in compiled rules

### ISSUE THREAD

**Issue by EthanEChristian** [cerner/clara-rules#402]
When clara creates [join-filter-fns](https://github.com/cerner/clara-rules/blob/master/src/main/clojure/clara/rules/compiler.clj#L460) it typically creates a let block with assignments. These assignments are extracted from the token's bindings via `get-in`, using a simple tests:

``` Clojure
(ns clara.performance.test-fire-rules
  (:require [clojure.test :refer :all]
            [clara.rules.accumulators :as acc]
            [clara.rules :as r]
            [clara.rules.compiler :as com]))

<Performance test code written as part of #381 copied over>

(defrecord AFact [id])
(defrecord BFact [id])
(defrecord ParentFact [a-ids b-ids])

(def counter (atom {:a-count 0
                    :b-count 0}))

(deftest test-get-in-perf
  (let [accum-rule [{:ns-name (ns-name *ns*)
                     :lhs [{:type ParentFact
                            :constraints []
                            :fact-binding :?parent}
                           {:accumulator `(acc/all)
                            :from {:type AFact
                                   :constraints ['(contains? (:a-ids ?parent) id)]}
                            :result-binding :?as}
                           {:accumulator `(acc/all)
                            :from {:type BFact
                                   :constraints ['(contains? (:b-ids ?parent) id)]}
                            :result-binding :?bs}]
                     :rhs '(do (swap! counter update :a-count inc))}]

        parents (for [_ (range 5000)]
                  (->ParentFact
                    (set (range 400))
                    (set (range 400 800))))
        a-facts (for [id (range 5000)]
                  (->AFact id))
        b-facts (for [id (range 5000)]
                  (->BFact id))

        facts (doall (concat parents
                             a-facts
                             b-facts))

        session (com/mk-session [accum-rule])]
    (run-performance-test {:description "Slow get-in perf"
                           :func #(-> session
                                      (r/insert-all facts)
                                      (r/fire-rules))
                           :iterations 5
                           :mean-assertion (partial > 30000)})))
```

The rule in question in clara's dsl would look like:
``` Clojure
(r/defrule test-rule
  [?parent <- ParentFact]
  [?as <- (acc/all) :from [AFact (contains? (:a-ids ?parent) id)]]
  [?bs <- (acc/all) :from [BFact (contains? (:b-ids ?parent) id)]]
  =>
  (swap! counter update :a-count inc))
```

While the test is an unlikely example, it is a bit slower than i would expect. Running the performance test locally give me:
```
Running Performance tests for:
Slow get-in perf
==========================================
Mean: 24582.4ms
Standard Deviation: 1256.9419397887875ms
```

Gathering [VisualVM Snapshot](https://github.com/cerner/clara-rules/files/2214851/Clara_get-in_perf.nps.zip) shows that a fair portion of the fire-rules time is dedicated to `get-in`, even more time than the set `contains?` call.

With some quick changes similar to:
``` Clojure
?parent (get-in ?__token__ [:bindings :?parent])
```
to 
``` Clojure
?parent (:?parent (:bindings ?__token__))
```
Drops the performance test above to:
```
Running Performance tests for:
Slow get-in perf
==========================================
Mean: 16982.2ms
Standard Deviation: 963.0259394221945ms
```

Like mentioned above this scenario is probably unlikely, however in larger systems where rules are fired a lot, this time will likely add up significantly.

**Comment by EthanEChristian** [comment]
I am hoping on getting a PR up this weekend to change this as it seems like pretty low hanging fruit.

**Comment by mrrodriguez** [comment]
@WilliamParker 

> Quick thought: if you constrain the lookup in get-in to be a vector of keywords known at compile time, it would be fairly easy to create a macro version that would expand to one at a time lookups.


Couldn't you just generated a call to the threading macro?
```clj

(-> x :a :b :c)
```

**Comment by mrrodriguez** [comment]
This is an interesting find. I've also found `get-in` to be surprisingly slow before when the same can be accomplished statically via direct lookups. There are several factors involved I'm sure. Along with the general overhead of walking through the keyseq dynamically, there is also a loss of performance around keyword direct invocation callsite caching.

**Comment by alex-dixon** [comment]
@EthanEChristian Really interesting. Thanks for your work on this.

**Comment by WilliamParker** [comment]
Regarding "Couldn't you just generated a call to the threading macro?"

Yes, you can since these are all keywords implementing ILookup.  I see that is

**Comment by EthanEChristian** [comment]
I have merged #404, and will be closing this issue.
